Sources/Core/Microsoft.StreamProcessing/Operators/SnapshotWindow/Tumbling/SnapshotWindowTumblingTemplate.tt

```diff
@@ -205,116 +185,77 @@ internal sealed class <#= className #><#= genericParameters #> : UnaryPipe<<#= T
+<# } else { #>
+                int iter1 = FastDictionary<<#= TKey #>, HeldState>.IteratorStart;
+                while (this.heldAggregates.Iterate(ref iter1))
+                {
+                    var iter1entry = this.heldAggregates.entries[iter1];
+
+                    int c = this.batch.Count;
+                    this.batch.vsync.col[c] = this.lastSyncTime;
+                    this.batch.vother.col[c] = this.lastSyncTime + this.hop;
+                    <#= assignToOutput(computeResult("iter1entry.value")) #>
+                    this.batch.key.col[c] = iter1entry.key;
+                    this.batch.hash.col[c] = this.keyComparerGetHashCode(iter1entry.key);
+                    this.batch.Count++;
+                    if (this.batch.Count == Config.DataBatchSize) FlushContents();
+                }
+
+                // Time has moved forward, clean the held aggregates
+                this.heldAggregates.Clear();
+<# } #>
 
                 // Since sync time changed, set this.lastSyncTime
                 this.lastSyncTime = syncTime;
             }
 
-            <# if (this.isUngrouped) { #>
-
+<# if (this.isUngrouped) { #>
             if (currentState == null)
             {
                 currentState = new HeldState();
                 currentState.state.state = <#= initialState #>;
-                currentState.state.active = 1; // start edge only
                 currentState.timestamp = syncTime;
             }
-            <# } else { #>
-
-            if (currentState == null || currentHash != col_hash[i] || !<#= inlinedKeyComparerEquals("currentKey", "col_key[i]") #>)
-            {
-                // Need to retrieve the key from the dictionary
-                currentKey = col_key[i];
-                currentHash = col_hash[i];
-
-                int index;
-
-                if (!this.heldAggregates.Lookup(currentKey, currentHash, out index))
-                {
-                    // New group. Create new state
-                    currentState = new HeldState();
-                    currentState.state.state = <#= initialState #>;
-                    currentState.timestamp = syncTime;
-                    // No output because initial state is empty
-
-                    this.heldAggregates.Insert(ref index, currentKey, currentState);
-                }
-                else
-                {
-                    // read new currentState from _heldAgg index
-                    currentState = this.heldAggregates.entries[index].value;
-                }
-            }
-            else
-            {
-                if (syncTime > currentState.timestamp)
-                {
-                    // Reset currentState
-                    currentState.state.state = <#= initialState #>;
-                    currentState.timestamp = syncTime;
-                }
-            }
-            <# } #>
 
             currentState.state.state = <#= accumulate("currentState.state.state", "col_vsync[i]") #>;
```

Comment:
>currentState.state.state [](start = 12, length = 24)

This amount of indirection is confusing. Is currentState.state assumed to be of type StateAndActive? Is that guaranteed to be true?

Reply:
Note that this isn't a change but a code move - but yes, that's guaranteed to be true. The nice part is that this line is only called on the "simple" case (ungrouped) now instead of all the time.